scipy/sparse/tests/test_construct.py

```diff
@@ -340,8 +340,8 @@ def test_kronsum(self):
 
     def test_vstack(self):
```

Comment:
I've been warned that the combinatorics of the sparse subpackage makes testing expensive.  So, I'd like to avoid duplication of testing where possible.
Here, we test the base code by testing the arrays interface. And then we test the spmatrix aspects of the code. So I think the coverage is the same as your suggestion of more tests.

Reply:
> I've been warned that the combinatorics of the sparse subpackage makes testing expensive. So, I'd like to avoid duplication of testing where possible.

I think this can be true, but not sure this case will have much effect. Trying to time this test:

<details>
<summary> Full output </summary>

(scipy-dev) isaac@Mimir:~/github/scipy ‹block_stack›
$ python dev.py test -s sparse  -- -k test_vstack --durations=0 -vv
💻  ninja -C /Users/isaac/github/scipy/build -j8
ninja: Entering directory `/Users/isaac/github/scipy/build'
[4/4] Generating scipy/generate-version with a custom command
Build OK
💻  meson install -C build --only-changed
Installing, see meson-install.log...
Installation OK
SciPy from development installed path at: /Users/isaac/github/scipy/build-install/lib/python3.10/site-packages
Running tests for scipy version:1.12.0.dev0+1321.c969aa1, installed at:/Users/isaac/github/scipy/build-install/lib/python3.10/site-packages/scipy
============================================================================ test session starts ============================================================================
platform darwin -- Python 3.10.11, pytest-7.3.1, pluggy-1.0.0 -- /Users/isaac/miniconda3/envs/scipy-dev/bin/python3
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/Users/isaac/github/scipy/build-install/lib/python3.10/site-packages/.hypothesis/examples')
metadata: {'Python': '3.10.11', 'Platform': 'macOS-13.4.1-x86_64-i386-64bit', 'Packages': {'pytest': '7.3.1', 'pluggy': '1.0.0'}, 'Plugins': {'timeout': '2.1.0', 'json-report': '1.5.0', 'xdist': '3.3.1', 'cov': '4.0.0', 'hypothesis': '6.81.2', 'metadata': '2.0.4'}}
rootdir: /Users/isaac/github/scipy
configfile: pytest.ini
plugins: timeout-2.1.0, json-report-1.5.0, xdist-3.3.1, cov-4.0.0, hypothesis-6.81.2, metadata-2.0.4
collected 10535 items / 10534 deselected / 1 selected                                                                                                                       

scipy/sparse/tests/test_construct.py::TestConstructUtils::test_vstack PASSED                                                                                          [100%]

============================================================================= slowest durations =============================================================================
0.00s setup    build-install/lib/python3.10/site-packages/scipy/sparse/tests/test_construct.py::TestConstructUtils::test_vstack
0.00s call     build-install/lib/python3.10/site-packages/scipy/sparse/tests/test_construct.py::TestConstructUtils::test_vstack
0.00s teardown build-install/lib/python3.10/site-packages/scipy/sparse/tests/test_construct.py::TestConstructUtils::test_vstack
==================================================================== 1 passed, 10534 deselected in 1.62s ====================================================================

</details>

```
=============================================================================
0.00s setup    build-install/lib/python3.10/site-packages/scipy/sparse/tests/test_construct.py::TestConstructUtils::test_vstack
0.00s call     build-install/lib/python3.10/site-packages/scipy/sparse/tests/test_construct.py::TestConstructUtils::test_vstack
0.00s teardown build-install/lib/python3.10/site-packages/scipy/sparse/tests/test_construct.py::TestConstructUtils::test_vstack
```

> Here, we test the base code by testing the arrays interface. And then we test the spmatrix aspects of the code. So I think the coverage is the same as your suggestion of more tests.

Yes, but this assumes the underlying codepath for `vstack` on matrices is the same as arrays, and that the wiring to get to that codepath is correct. The changes here would ensure correctness without having to rely on that assumption.

Broadly, my idea here is "testing the interface instead of the implementation"

Reply:
I've parametrized the vstack and hstack tests to run for both matrix and array classes. I separated the tests of return matrix/array into a separate test method.